addons/crm/models/crm_lead.py

```diff
@@ -1468,18 +1468,38 @@ def _cron_update_automated_probabilities(self):
             leads_to_update = self.env['crm.lead'].search(pending_lead_domain)
             lead_probabilities = leads_to_update._pls_get_naive_bayes_probabilities(batch_mode=True)
 
-            # Update in execute batch to avoid server roundtrips + page_size to 10000 to avoid memory errors
+            # Group by new probability to reduce server roundtrips when executing the update
+            lead_ids_by_probabilities = defaultdict(list)
+            for lead_id, proba in sorted(lead_probabilities.items()):
+                lead_ids_by_probabilities[proba].append(lead_id)
+
             # Update both probability and automated_probability if they were equal, else, update only automated_probability
             sql = """UPDATE crm_lead
                         SET automated_probability = %s,
-                            probability = CASE WHEN (ROUND(probability::numeric, 2) = ROUND(automated_probability::numeric, 2) or probability is null)
+                            probability = CASE WHEN (probability = automated_probability or probability is null)
                                                THEN (%s)
                                                ELSE (probability)
                                                END
-                        WHERE id = %s"""
-            batch_params = [(lead_probabilities[lead.id], lead_probabilities[lead.id], lead.id) for lead in leads_to_update if lead.id in lead_probabilities]
-            extras.execute_batch(self._cr, sql, batch_params, page_size=10000)
-            _logger.info("Predictive Lead Scoring : all automated probability updated (count: %d)" % (len(leads_to_update)))
+                        WHERE id in %s"""
+
+            # Update by a limited number of leads at the same time :
+            # - avoid memory errors
+            # - avoid to block the table for too long with a too big transaction
+            step = 5
+            updated_leads = 0
+            nbr_of_transactions = 0
+            nbr_leads_to_update = len(leads_to_update)
+            _logger.info("Predictive Lead Scoring : Start updating automated probability (%d records)" % nbr_leads_to_update)
+            for proba, lead_ids in lead_ids_by_probabilities.items():
+                for idx_from in range(0, len(lead_ids), step):
+                    lead_ids_to_exec = lead_ids[idx_from:] if idx_from + step >= len(lead_ids) else lead_ids[idx_from:idx_from + step]
```

Comment:
```suggestion
                    lead_ids_to_exec = lead_ids[idx_from:idx_from + step]
```
You can do it this way, it won't crash and will stop at the last element of the list